modules/bullet/space_bullet.cpp

```diff
@@ -890,202 +890,150 @@ void SpaceBullet::update_gravity() {
+			}
 
-			if (has_penetration) {
+			KinematicCollisionResult collision_result;
+			if (kinematic_collision_test(p_body, collision_transform, p_infinite_inertia, discard_motion, &collision_result)) {
 
-				const btRigidBody *btRigid = static_cast<const btRigidBody *>(r_recover_result.other_collision_object);
+				// Extract collision information.
+				const btRigidBody *btRigid = static_cast<const btRigidBody *>(collision_result.other_collision_object);
 				CollisionObjectBullet *collisionObject = static_cast<CollisionObjectBullet *>(btRigid->getUserPointer());
 
-				B_TO_G(motion, r_result->remainder); // is the remaining movements
-				r_result->remainder = p_motion - r_result->remainder;
-
-				B_TO_G(r_recover_result.pointWorld, r_result->collision_point);
-				B_TO_G(r_recover_result.normal, r_result->collision_normal);
-				B_TO_G(btRigid->getVelocityInLocalPoint(r_recover_result.pointWorld - btRigid->getWorldTransform().getOrigin()), r_result->collider_velocity); // It calculates velocity at point and assign it using special function Bullet_to_Godot
-				r_result->collider = collisionObject->get_self();
-				r_result->collider_id = collisionObject->get_instance_id();
-				r_result->collider_shape = r_recover_result.other_compound_shape_index;
-				r_result->collision_local_shape = r_recover_result.local_shape_most_recovered;
-
-#if debug_test_motion
-				Vector3 sup_line2;
-				B_TO_G(motion, sup_line2);
-				normalLine->clear();
-				normalLine->begin(Mesh::PRIMITIVE_LINES, NULL);
-				normalLine->add_vertex(r_result->collision_point);
-				normalLine->add_vertex(r_result->collision_point + r_result->collision_normal * 10);
-				normalLine->end();
-#endif
-			} else {
-				r_result->remainder = Vector3();
+				B_TO_G(collision_result.pointWorld, r_motion_result->collision_point);
+				B_TO_G(collision_result.normal, r_motion_result->collision_normal);
+				B_TO_G(btRigid->getVelocityInLocalPoint(collision_result.pointWorld - btRigid->getWorldTransform().getOrigin()), r_motion_result->collider_velocity);
+				r_motion_result->collider = collisionObject->get_self();
+				r_motion_result->collider_id = collisionObject->get_instance_id();
+				r_motion_result->collider_shape = collision_result.other_compound_shape_index;
+				r_motion_result->collision_local_shape = collision_result.local_shape_most_recovered;
+
+				// Use binary search to find the safe motion.
+				btScalar safe = 0.f;
+				btScalar unsafe = 1.f;
+				for (int step = 0; step < 8; step++) {
+
+					btScalar mid_point = (safe + unsafe) * 0.5f;
+					btTransform test_transform = body_transform;
+					test_transform.getOrigin() += motion * mid_point;
+					if (kinematic_collision_test(p_body, test_transform, p_infinite_inertia, discard_motion)) {
+						unsafe = mid_point;
+					} else {
+						safe = mid_point;
+					}
+				}
+				motion *= safe;
 			}
 		}
+
+		B_TO_G(motion + recover_motion, r_motion_result->motion);
+		B_TO_G(motion, r_motion_result->remainder);
```

Comment:
This line has no effect and should be removed.

Reply:
Good spot. This is a remnant of the old code. I'll remove it.